include/dba.php

```diff
@@ -662,17 +682,35 @@ static public function e($sql) {
 
 		$args = func_get_args();
 
-		$stmt = call_user_func_array('self::p', $args);
+		// In a case of a deadlock we are repeating the query 20 times
+		$timeout = 20;
 
-		if (is_bool($stmt)) {
-			$retval = $stmt;
-		} elseif (is_object($stmt)) {
-			$retval = true;
-		} else {
-			$retval = false;
-		}
+		do {
+			$stmt = call_user_func_array('self::p', $args);
 
-		self::close($stmt);
+			if (is_bool($stmt)) {
+				$retval = $stmt;
+			} elseif (is_object($stmt)) {
+				$retval = true;
+			} else {
+				$retval = false;
+			}
+
+			self::close($stmt);
+
+		} while ((self::$dbo->errorno == 1213) && (--$timeout > 0));
```

Comment:
Plus invert the comparison:
`} while ((1213 == self::$dbo->errorno) && (--$timeout > 0));`
This avoids accidently assigning values:
`} while ((self::$dbo->errorno = 1213) && (--$timeout > 0));`
Very hard to notice, but this is very easy:
`} while ((1213 = self::$dbo->errorno ) && (--$timeout > 0));`
As PHP will throw an error at you.